File: src/Blokemon.CardGen/Domain/Card.cs
using System.Collections.Immutable;

namespace Blokemon.CardGen.Domain;

/// <summary>One card face.</summary>
public interface ICard
{
    /// <summary>The canonical identifier of the card.</summary>
    CardId Id { get; }

    /// <summary>The regions printed on the face, in print order.</summary>
    ImmutableArray<CardRegion> Regions { get; }

    /// <summary>The stylesheet token of the card's atmosphere, absent on unprinted stock.</summary>
    string? ThemeToken { get; }

    /// <summary>The type printed on the card, absent when the card has no type disc.</summary>
    BlokemonType? PrintedType { get; }

    /// <summary>The name printed on the card, falling back to its identifier.</summary>
    string DisplayName { get; }
}

/// <summary>One card face printed in a given atmosphere.</summary>
/// <typeparam name="TTheme">The atmosphere the card prints in.</typeparam>
/// <param name="Id">The canonical identifier of the card.</param>
/// <param name="Regions">The regions printed on the face, in print order.</param>
public sealed record Card<TTheme>(CardId Id, ImmutableArray<CardRegion> Regions) : ICard
    where TTheme : ICardTheme
{
    /// <inheritdoc/>
    public string? ThemeToken => TTheme.Token;

    /// <inheritdoc/>
    public BlokemonType? PrintedType =>
        Regions
            .OfType<CardRegion.Vitality>()
            .Select(vitality => (BlokemonType?)vitality.Type)
            .FirstOrDefault();

    /// <inheritdoc/>
    public string DisplayName =>
        Regions
            .OfType<CardRegion.Nameplate>()
            .Select(nameplate => nameplate.Name)
            .DefaultIfEmpty(Id.Value)
            .First();
}

File: src/Blokemon.CardGen/Domain/Card.fs
namespace Blokemon.CardGen.Domain

open System.Collections.Immutable

/// One card face printed in a given atmosphere.
type Card =
    {
        /// The canonical identifier of the card.
        Id: CardId

        /// The regions printed on the face, in print order.
        Regions: ImmutableArray<CardRegion>

        /// The stylesheet token of the card's atmosphere, absent on unprinted stock.
        ThemeToken: string option
    }

    /// The type printed on the card, absent when the card has no type disc.
    member this.PrintedType =
        this.Regions
        |> Seq.tryPick (function
            | CardRegion.Vitality(printedType = printedType) -> Some printedType
            | _ -> None)

    /// The name printed on the card, falling back to its identifier.
    member this.DisplayName =
        this.Regions
        |> Seq.tryPick (function
            | CardRegion.Nameplate(name = name) -> Some name
            | _ -> None)
        |> Option.defaultValue this.Id.Value
